File: app/src/main/java/ir/ghararemaghzha/game/fragments/VerifyFragment.java
package ir.ghararemaghzha.game.fragments;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;

import com.google.android.gms.auth.api.phone.SmsRetriever;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.messaging.FirebaseMessaging;

import io.realm.Realm;
import ir.ghararemaghzha.game.R;
import ir.ghararemaghzha.game.activities.MainActivity;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.data.RetrofitClient;
import ir.ghararemaghzha.game.dialogs.GetDataDialog;
import ir.ghararemaghzha.game.models.GeneralResponse;
import ir.ghararemaghzha.game.models.QuestionModel;
import ir.ghararemaghzha.game.models.QuestionResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static ir.ghararemaghzha.game.classes.Const.FCM_TOPIC;
import static ir.ghararemaghzha.game.classes.Utils.convertToTimeFormat;


public class VerifyFragment extends Fragment {

    private Context context;
    private FragmentActivity activity;
    private MaterialTextView resend;
    private TextInputEditText code;
    private MaterialButton verify;
    private CountDownTimer timer;
    private long timerTime = 120000;

    private String userName;
    private String accessToken;
    private String number;
    private String userId;

    private GetDataDialog dialog;

    public VerifyFragment() {
    }

    private BroadcastReceiver rec = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getExtras() != null) {
                String c = intent.getExtras().getString("code");
                if (c != null && c.length() == 6) {
                    code.setText(c);

                }
            }
        }
    };

    public void setNumber(String number) {
        this.number = number;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_verify, container, false);
        context = getContext();
        activity = getActivity();

        context.registerReceiver(rec, new IntentFilter("codeReceived"));
        init(v);

        return v;
    }

    private void init(View v) {
        resend = v.findViewById(R.id.verify_resend);
        resend.setEnabled(false);
        code = v.findViewById(R.id.verify_code);
        verify = v.findViewById(R.id.verify_verify);
        SmsRetriever.getClient(activity).startSmsUserConsent("98300077");
        initTimer();
        onClicks();

        code.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (charSequence.length() == 6)
                    Utils.hideKeyboard(activity);
            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() == 6)
                    Utils.hideKeyboard(activity);
            }
        });

    }

    private void onClicks() {
        resend.setOnClickListener(v -> {
            doResend();
        });
        verify.setOnClickListener(v -> {
            String c = code.getText().toString();
            if (c.length() < 6 || c.startsWith("0")) {
                Toast.makeText(context, context.getString(R.string.verify_wrong_code), Toast.LENGTH_SHORT).show();
            } else {
                if (Utils.checkInternet(context))
                    doVerify(c);
                else
                    Toast.makeText(context, context.getString(R.string.general_internet_error), Toast.LENGTH_SHORT).show();
            }


        });


    }

    private void initTimer() {
        timer = new CountDownTimer(timerTime, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                resend.setText(convertToTimeFormat(millisUntilFinished));
            }

            @Override
            public void onFinish() {
                try {
                    resend.setEnabled(true);
                    resend.setText(getString(R.string.verify_resend));
                } catch (Exception e) {
                    e.printStackTrace();
                }

            }
        };
        timer.start();
    }

    private void doResend() {
        RetrofitClient.getInstance().getApi()
                .resend(number)
                .enqueue(new Callback<GeneralResponse>() {
                    @Override
                    public void onResponse(Call<GeneralResponse> call, Response<GeneralResponse> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {
                            Toast.makeText(context, context.getString(R.string.general_send), Toast.LENGTH_SHORT).show();
                            timerTime *= 2;
                            timer.cancel();
                            initTimer();

                        } else {
                            Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<GeneralResponse> call, Throwable t) {
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();

                    }
                });
    }

    private void doVerify(String code) {
        Utils.hideKeyboard(activity);
        verify.setEnabled(false);
        verify.setText("...");
        String fbToken = Utils.getFbToken(context);
        RetrofitClient.getInstance().getApi()
                .verification(number, code, fbToken)
                .enqueue(new Callback<GeneralResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<GeneralResponse> call, @NonNull Response<GeneralResponse> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {
                            FirebaseMessaging.getInstance().subscribeToTopic(FCM_TOPIC);
                            userId = response.body().getUserId();
                            userName = response.body().getUserName();
                            accessToken = response.body().getToken();
                            String userCode = response.body().getUserCode();
                            String score = response.body().getUserScore();
                            String plan = response.body().getUserPlan();
                            MySharedPreference.getInstance(context).setNumber(number);
                            MySharedPreference.getInstance(context).setUsername(userName);
                            MySharedPreference.getInstance(context).setAccessToken(accessToken);
                            MySharedPreference.getInstance(context).setUserCode(userCode);
                            MySharedPreference.getInstance(context).setScore(score);
                            MySharedPreference.getInstance(context).setPlan(plan);
                            MySharedPreference.getInstance(context).setUserSex(response.body().getUserSex());
                            MySharedPreference.getInstance(context).setUserBday(response.body().getUserBday());
                            MySharedPreference.getInstance(context).setUserEmail(response.body().getUserEmail());
                            MySharedPreference.getInstance(context).setUserInvite(response.body().getUserInvite());
                            dialog = Utils.showGetDataLoading(context);
                            getQuestions();


                        } else if (response.code() == 401) {
                            verify.setEnabled(true);
                            verify.setText(context.getString(R.string.verify_verify));
                            Toast.makeText(context, context.getString(R.string.verify_wrong_code), Toast.LENGTH_SHORT).show();
                        } else {
                            verify.setEnabled(true);
                            verify.setText(context.getString(R.string.verify_verify));
                            Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<GeneralResponse> call, @NonNull Throwable t) {
                        verify.setEnabled(true);
                        verify.setText(context.getString(R.string.verify_verify));
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void getQuestions() {
        Realm db = Realm.getDefaultInstance();
        if (number.isEmpty() || accessToken.isEmpty()) {
            Utils.logout(activity,true);
            return;
        }
        RetrofitClient.getInstance().getApi()
                .getQuestions("Bearer " + accessToken, number, "0", "0")
                .enqueue(new Callback<QuestionResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<QuestionResponse> call, @NonNull Response<QuestionResponse> response) {
                        verify.setEnabled(true);
                        verify.setText(context.getString(R.string.verify_verify));
                        if (response.isSuccessful() && response.body() != null
                                && response.body().getResult().equals("success")
                                && !response.body().getMessage().equals("empty")) {
                            for (QuestionModel model : response.body().getData()) {
                                if (model.getUserAnswer().equals("-1")) {
                                    model.setUploaded(false);
                                } else {
                                    model.setUploaded(true);
                                }
                                model.setVisible(false);
                                db.executeTransaction(realm1 -> realm1.insertOrUpdate(model));
                            }
                            db.close();


                            MySharedPreference.getInstance(context).setUserId(userId);
                            Toast.makeText(context, context.getString(R.string.verify_welcome, userName), Toast.LENGTH_SHORT).show();
                            dialog.dismiss();
                            startActivity(new Intent(activity, MainActivity.class));
                            activity.overridePendingTransition(R.anim.enter_right, R.anim.exit_left);
                            activity.finish();
                        } else {
                            Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                            dialog.dismiss();

                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<QuestionResponse> call, @NonNull Throwable t) {
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                        dialog.dismiss();
                        verify.setEnabled(true);
                        verify.setText(context.getString(R.string.verify_verify));
                    }
                });


    }


    @Override
    public void onDestroy() {
        context.unregisterReceiver(rec);
        super.onDestroy();
    }
}

File: app/src/main/java/ir/ghararemaghzha/game/fragments/VerifyFragment.kt
package ir.ghararemaghzha.game.fragments

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.core.widget.doAfterTextChanged
import androidx.core.widget.doOnTextChanged
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.navigation.findNavController
import com.google.android.gms.auth.api.phone.SmsRetriever
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textview.MaterialTextView
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.messaging.FirebaseMessaging
import io.realm.Realm
import ir.ghararemaghzha.game.R
import ir.ghararemaghzha.game.classes.Const.FCM_TOPIC
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.data.ApiRepository
import ir.ghararemaghzha.game.data.NetworkApi
import ir.ghararemaghzha.game.data.RemoteDataSource
import ir.ghararemaghzha.game.data.Resource
import ir.ghararemaghzha.game.dialogs.GetDataDialog
import ir.ghararemaghzha.game.models.QuestionModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class VerifyFragment : Fragment(R.layout.fragment_verify) {
    private lateinit var mFirebaseAnalytics: FirebaseAnalytics
    private lateinit var resend: MaterialTextView
    private lateinit var code: TextInputEditText
    private lateinit var verify: MaterialButton
    private lateinit var timer: CountDownTimer
    private var timerTime = 120000L
    private lateinit var userName: String
    private lateinit var accessToken: String
    private var number: String = ""
    private lateinit var userId: String
    private lateinit var dialog: GetDataDialog
    private lateinit var act: FragmentActivity
    private lateinit var ctx: Context

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v= super.onCreateView(inflater, container, savedInstanceState)
        act=requireActivity()
        ctx = requireContext()
        return v
    }

    private val rec = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val c = intent?.extras?.getString("code")
            if (c?.length == 6) {
                code.setText(c)
                if (Utils.checkInternet(ctx)) {
                    verify.isEnabled = false
                    verify.text = "..."
                    Utils.hideKeyboard(act)
                    CoroutineScope(Dispatchers.IO).launch {
                        doVerify(c)
                    }
                } else
                    Toast.makeText(ctx, R.string.general_internet_error, Toast.LENGTH_SHORT).show()

            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(ctx)

        if (arguments != null)
            number = requireArguments().getString("number", "")
       ctx.registerReceiver(rec, IntentFilter("codeReceived"))
        init(view)

        val callback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {}
        }
        act.onBackPressedDispatcher.addCallback(viewLifecycleOwner, callback)
    }


    private fun init(v: View) {
        resend = v.findViewById(R.id.verify_resend)
        resend.isEnabled = false
        code = v.findViewById(R.id.verify_code)
        verify = v.findViewById(R.id.verify_verify)
        SmsRetriever.getClient(act).startSmsUserConsent(null)
        initTimer()
        onClicks()

        code.doAfterTextChanged {
            if (it?.length == 6) Utils.hideKeyboard(act)
        }
        code.doOnTextChanged { s, _, _, _ ->
            if (s?.length == 6) Utils.hideKeyboard(act)
        }
    }

    private fun onClicks() {
        resend.setOnClickListener { CoroutineScope(Dispatchers.IO).launch { doResend() } }
        verify.setOnClickListener {
            val c = code.text.toString()
            if (c.length < 6 || c.startsWith("0")) {
                Toast.makeText(ctx, R.string.verify_wrong_code, Toast.LENGTH_SHORT).show()
            } else {
                if (Utils.checkInternet(ctx)) {
                    Utils.hideKeyboard(act)
                    verify.isEnabled = false
                    verify.text = "..."
                    CoroutineScope(Dispatchers.IO).launch {
                        doVerify(c)
                    }
                } else
                    Toast.makeText(ctx, R.string.general_internet_error, Toast.LENGTH_SHORT).show()
            }
        }

    }

    private fun initTimer() {
        timer = object : CountDownTimer(timerTime, 1000L) {
            override fun onTick(millisUntilFinished: Long) {
                resend.text = Utils.convertToTimeFormat(millisUntilFinished)
            }

            override fun onFinish() {
                resend.isEnabled = true
                resend.setText(R.string.verify_resend)
            }
        }
        timer.start()
    }

    private suspend fun doResend() {
        when (val res = ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java))
                .resend(number)) {
            is Resource.Success -> {
                withContext(Dispatchers.Main) {
                    Toast.makeText(ctx, R.string.general_send, Toast.LENGTH_SHORT).show()
                    timerTime *= 2
                    timer.cancel()
                    initTimer()
                }
            }
            is Resource.Failure -> {
                if (res.isNetworkError) {
                    withContext(Dispatchers.Main) {
                        Toast.makeText(ctx, R.string.general_internet_error, Toast.LENGTH_SHORT).show()
                    }

                } else {
                    withContext(Dispatchers.Main) {
                        Toast.makeText(ctx, R.string.general_error, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    private suspend fun doVerify(code: String) {
        val fbToken = Utils.getFbToken(ctx)
        when (val res = ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java)).verification(number, code, fbToken)) {
            is Resource.Success -> {
                FirebaseMessaging.getInstance().subscribeToTopic(FCM_TOPIC)

                userId = res.value.userId
                userName = res.value.userName
                accessToken = res.value.token
                val userCode = res.value.userCode
                val score = res.value.userScore.toInt()

                MySharedPreference.getInstance(ctx).setNumber(number)
                MySharedPreference.getInstance(ctx).setUsername(userName)
                MySharedPreference.getInstance(ctx).setAccessToken(accessToken)
                MySharedPreference.getInstance(ctx).setUserCode(userCode)
                MySharedPreference.getInstance(ctx).setScore(score)

                MySharedPreference.getInstance(ctx).setUserSex(res.value.userSex ?: "")
                MySharedPreference.getInstance(ctx).setUserBirthday(res.value.userBday ?: "")
                MySharedPreference.getInstance(ctx).setUserEmail(res.value.userEmail ?: "")
                MySharedPreference.getInstance(ctx).setUserInvite(res.value.userInvite ?: "")
                if (!res.value.userAvatar.isNullOrEmpty())
                    MySharedPreference.getInstance(ctx).setUserAvatar(res.value.userAvatar)

                withContext(Dispatchers.Main) {
                    dialog = Utils.showGetDataLoading(ctx)
                }
                getQuestions()


            }
            is Resource.Failure -> {
                withContext(Dispatchers.Main) {
                    verify.isEnabled = true
                    verify.setText(R.string.verify_verify)
                    if (res.isNetworkError) {
                        Toast.makeText(ctx, R.string.general_internet_error, Toast.LENGTH_SHORT).show()
                    } else {
                        when (res.errorCode) {
                            401 -> {
                                Toast.makeText(ctx, R.string.verify_wrong_code, Toast.LENGTH_SHORT).show()
                            }
                            else -> {
                                Toast.makeText(ctx, R.string.general_error, Toast.LENGTH_SHORT).show()
                            }
                        }
                    }
                }
            }
        }
    }

    private suspend fun getQuestions() {
        when (val res = ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java)).getQuestions("Bearer $accessToken", number)) {

            is Resource.Success -> {
                withContext(Dispatchers.Main) {
                    if (res.value.message != "empty") {
                        val data: MutableCollection<QuestionModel> = mutableListOf()
                        for ( model in res.value.data) {
                            model.uploaded = model.userAnswer != "-1"
                            data.add(model)
                        }
                        val db = Realm.getDefaultInstance()
                        db.executeTransaction { it.insertOrUpdate(data) }
                        MySharedPreference.getInstance(ctx).setUserId(userId)
                        Toast.makeText(ctx, getString(R.string.verify_welcome, userName), Toast.LENGTH_SHORT).show()
                        logEvent()
                        dialog.dismiss()
                        view?.findNavController()?.navigate(R.id.action_verifyFragment_to_slidesActivity)
                        act.finish()
                    } else {
                        verify.isEnabled = true
                        verify.setText(R.string.verify_verify)
                        dialog.dismiss()
                        Toast.makeText(ctx, R.string.general_error, Toast.LENGTH_SHORT).show()
                    }
                }


            }
            is Resource.Failure -> {
                withContext(Dispatchers.Main) {
                    verify.isEnabled = true
                    verify.setText(R.string.verify_verify)
                    dialog.dismiss()
                    if (res.isNetworkError)
                        Toast.makeText(ctx, R.string.general_internet_error, Toast.LENGTH_SHORT).show()
                    else
                        Toast.makeText(ctx, R.string.general_error, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        ctx.unregisterReceiver(rec)
    }

    private fun logEvent() {
        val bundle = Bundle()
        bundle.putString(FirebaseAnalytics.Param.ITEM_ID, userId)
        bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, userName)
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN, bundle)
    }
}